### PR TITLE
Fix double localizing in `consequence_for_input`

### LINKED ISSUE #16128: panicked at 'Datum::unwrap_float64 called on Int64(0)', src/expr/src/scalar/func.rs:292:15 with aggregate inside a derived table
### What version of Materialize are you using?

v0.33.0-dev (5ff0dacee)

### How did you install Materialize?

Docker image

### What is the issue?

This panics:

```
CREATE TABLE t1 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);

CREATE TABLE t2 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);

EXPLAIN SELECT *
FROM t2 , 
(
    SELECT AVG(1) AS f1 , COUNT (1) AS f2
    FROM t2
    WHERE f2 IS NULL
) AS a2
WHERE a2.f1 = 1
OR a2.f1 + 4 = 8
AND t2.f2 = 1;
```

~~I am not able to reproduce it with other aggregates except AVG . Can it be because AVG gets decomposed into a SUM / COUNT?~~

### Relevant log output

```source
thread 'coordinator' panicked at 'Datum::unwrap_float64 called on Int64(0)', src/expr/src/scalar/func.rs:292:15
stack backtrace:
   0: rust_begin_unwind
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/897e37553bba8b42751c67658967889d11ecd120/library/core/src/panicking.rs:142:14
   2: mz_repr::scalar::Datum::unwrap_float64
   3: mz_expr::scalar::func::add_float64
   4: mz_expr::scalar::func::BinaryFunc::eval
   5: mz_expr::scalar::MirScalarExpr::eval
   6: mz_expr::scalar::func::BinaryFunc::eval
   7: mz_expr::scalar::MirScalarExpr::eval
   8: mz_expr::scalar::func::VariadicFunc::eval
   9: mz_expr::scalar::MirScalarExpr::eval
  10: mz_transform::reduction::FoldConstants::fold_filter_constant
  11: mz_transform::reduction::FoldConstants::action
  12: mz_expr::visit::StackSafeVisit<T>::try_visit_mut_post
  13: <mz_expr::relation::MirRelationExpr as mz_expr::visit::VisitChildren<mz_expr::relation::MirRelationExpr>>::try_visit_mut_children
  14: mz_expr::visit::StackSafeVisit<T>::try_visit_mut_post
  15: <mz_expr::relation::MirRelationExpr as mz_expr::visit::VisitChildren<mz_expr::relation::MirRelationExpr>>::try_visit_mut_children
  16: mz_expr::visit::StackSafeVisit<T>::try_visit_mut_post
  17: <mz_expr::relation::MirRelationExpr as mz_expr::visit::VisitChildren<mz_expr::relation::MirRelationExpr>>::try_visit_mut_children
  18: mz_expr::visit::StackSafeVisit<T>::try_visit_mut_post
  19: <mz_transform::reduction::FoldConstants as mz_transform::Transform>::transform
  20: <mz_transform::FuseAndCollapse as mz_transform::Transform>::transform
  21: tracing::span::Span::in_scope
  22: <mz_transform::Fixpoint as mz_transform::Transform>::transform
  23: mz_transform::Optimizer::optimize
  24: tracing_core::dispatcher::with_default
  25: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator<S>>::sequence_explain
  26: mz_adapter::coord::sequencer::<impl mz_adapter::coord::Coordinator<S>>::sequence_plan::{{closure}}::{{closure}}
  27: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  28: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  29: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  30: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  31: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  32: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  33: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  34: mz_adapter::coord::message_handler::<impl mz_adapter::coord::Coordinator<S>>::handle_message::{{closure}}
  35: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  36: tokio::park::thread::CachedParkThread::block_on
  37: tokio::runtime::handle::Handle::block_on
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

### PR DESCRIPTION
**Pull request by ggevay**
`JoinInputMapper::consequence_for_input` was calling `map_expr_to_local` too many times: If the recursion digs out a subexpression from somewhere deep inside an expression, then we were calling `map_expr_to_local` at every level of the recursion. These calls were on overlapping subexpressions, so calls after the first call were sometimes trying to localize an already (partially) localized subexpression.

I removed all calls to `map_expr_to_local` from inside the recursion, and added a new call outside the recursion.  

(The bug was not noticed before, because it is not so easy to trigger: the first `map_expr_to_local` call often makes the index small enough that any subsequent call is a no-op, because the index is already on the first input if we interpret the index in the global context. To trigger the bug, the input that we are localizing into should have more fields than the first input, and the expression should be referring to one of the later columns. @philip-stoev, I'm wondering if you can maybe tweak something on your random SQL generator based on this experience, e.g., have more variation in the number of fields of different input tables.)

The randomly chosen reviewer is @antiguru.

### Motivation

  * This PR fixes a recognized bug. Fixes #16128

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
  - [ ] Added Philip's test from the issue and Frank's query from the other PR
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Fixed a bug in predicate pushdown.

### CODE REVIEW

**Comment by philip-stoev** [comment]
@frankmcsherry , @ggevay which of the two PRs is going to be the definitive one, so I can test it? (I will take feedback re. testing/grammar from both sides, thanks a lot for the pointers!)

**Comment by ggevay** [comment]
Modified the PR a bit, see https://github.com/MaterializeInc/materialize/pull/16185#issuecomment-1321863722

**Comment by frankmcsherry** [comment]
@philip-stoev This one has the same change as mine, and an additional test, so it seems better!

**Comment by ggevay** [comment]
(I carried over the M-release-blocker flag from https://github.com/MaterializeInc/materialize/pull/16185)

**Review by aalexandrov — APPROVED** [review]

**Comment by philip-stoev** [comment]
Item No 1. This branch is able to construct one extra CTE than the previous commit. I stared at the new plan but I was not able to figure out if the CTE is equivalent to the previous in-lined situation, given that the two locations that are now the same CTE were previously occupied by slightly different subtrees (apologies for butchering the terminology).

```
EXPLAIN SELECT
  FROM (
        SELECT f2
          FROM t2
       ) a1
  LEFT  JOIN t2
    ON TRUE
 WHERE 5 > a1 .f2
    OR f1 < 5
   AND a1 .f2 = 0 ;

--- 6875.out	2022-11-21 12:56:22.827778743 +0100
+++ 16875.out	2022-11-21 12:56:22.915780915 +0100
@@ -1,38 +1,38 @@
 Explained Query:
   Return
     Union
       Project ()
         Filter ((5 > #0) OR ((#0 = 0) AND (#1 < 5)))
           Get l0
       Project ()
         Join on=(#0 = #1) type=differential
           ArrangeBy keys=[[#0]]
             Union
               Negate
                 Distinct group_by=[#0]
                   Project (#0)
                     Filter (#2 OR #3) AND (#3 OR (#2 AND null))
                       Map ((#0 = 0), (5 > #0))
                         Get l0
               Distinct group_by=[#0]
-                Get l1
+                Project (#1)
+                  Filter (#2 OR #3) AND (#3 OR (#2 AND null))
+                    Map ((#1 = 0), (5 > #1))
+                      Get materialize.public.t2
           ArrangeBy keys=[[#0]]
-            Get l1
+            Project (#1)
+              Filter ((5 > #1) OR (null AND (#0 = 0) AND (#1 = 0)))
+                Get materialize.public.t2
   With
-    cte l1 =
-      Project (#1)
-        Filter (#2 OR #3) AND (#3 OR (#2 AND null))
-          Map ((#1 = 0), (5 > #1))
-            Get materialize.public.t2
     cte l0 =
       CrossJoin type=differential
         ArrangeBy keys=[[]]
           Project (#1)
             Get materialize.public.t2
         ArrangeBy keys=[[]]
           Project (#0)
             Get materialize.public.t2
 
 Used Indexes:
   - materialize.public.t2i1
```

**Comment by aalexandrov** [comment]
> Item No 1. 

What's the schema for that query?

**Comment by philip-stoev** [comment]
> > Item No 1.
> 
> What's the schema for that query?

Sorry, it is the same one from the SLT file. 

```
CREATE TABLE t1 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
CREATE TABLE t2 (f1 DOUBLE PRECISION, f2 DOUBLE PRECISION NOT NULL);
```

**Comment by philip-stoev** [comment]
Item No 2. In this branch, the opportunity to push/pull/hoist/lift? a predicate all the way into the source has been lost:

```
CREATE TABLE t1 (f2 INTEGER, f3 INTEGER);
CREATE TABLE t2 (f3 INTEGER,  f5 INTEGER);
CREATE TABLE t3 (f5 INTEGER, f6 INTEGER);
EXPLAIN SELECT *
  FROM t1
  JOIN (t2 LEFT JOIN t3 USING (f5)) USING (f3)
 WHERE f5 >= 1
    OR f6 <= f5 + 123456789
    AND f2 >= 0
  ;
```
--- 6875.out	2022-11-21 13:46:13.961110445 +0100
+++ 16875.out	2022-11-21 13:46:16.161164312 +0100
@@ -1,47 +1,45 @@
 Explained Query:
   Return
     Project (#1, #0, #3, #4)
       Filter ((#3 >= 1) OR ((#4 <= (#3 + 123456789)) AND (#0 >= 0)))
         Join on=(#1 = #2) type=differential
           ArrangeBy keys=[[#1]]
             Filter (#1) IS NOT NULL
               Get materialize.public.t1
           ArrangeBy keys=[[#0]]
             Union
-              Filter ((#2 <= (#1 + 123456789)) OR (#1 >= 1))
-                Get l0
+              Get l0
               Project (#0, #1, #4)
                 Map (null)
                   Join on=(#0 = #2 AND #1 = #3) type=differential
                     ArrangeBy keys=[[#0, #1]]
                       Union
                         Negate
                           Distinct group_by=[#0, #1]
                             Project (#0, #1)
-                              Filter (null OR (#1 >= 1))
-                                Get l0
+                              Get l0
                         Distinct group_by=[#0, #1]
                           Get l1
                     ArrangeBy keys=[[#0, #1]]
                       Get l1
   With
     cte l1 =
-      Filter (#0) IS NOT NULL AND (null OR (#1 >= 1))
+      Filter (#0) IS NOT NULL AND ((#1 <= (#1 + 123456789)) OR (#1 >= 1))
         Get materialize.public.t2
     cte l0 =
       Project (#0, #1, #3)
         Join on=(#1 = #2) type=differential
           ArrangeBy keys=[[#1]]
-            Filter (#0) IS NOT NULL AND (#1) IS NOT NULL
+            Filter (#0) IS NOT NULL AND (#1) IS NOT NULL AND ((#1 <= (#1 + 123456789)) OR (#1 >= 1))
               Get materialize.public.t2
           ArrangeBy keys=[[#0]]
-            Filter (#0) IS NOT NULL
+            Filter (#0) IS NOT NULL AND ((#0 <= (#0 + 123456789)) OR (#0 >= 1))
               Get materialize.public.t3
 
 Source materialize.public.t1
   filter=((#1) IS NOT NULL)
 Source materialize.public.t2
-  filter=((#0) IS NOT NULL)
+  filter=((#0) IS NOT NULL AND ((#1 <= (#1 + 123456789)) OR (#1 >= 1)))
 Source materialize.public.t3
-  filter=((#0) IS NOT NULL)
+  filter=((#0) IS NOT NULL AND ((#0 <= (#0 + 123456789)) OR (#0 >= 1)))
```

See the `f6 <= f5 + 123456789` predicate in particular

**Comment by philip-stoev** [comment]
Item No 3. This branch fixes this wrong-result situation. I guess this is expected, however the manifestation is different from the ones in the existing tests, so I would say it would be nice to push as an SLT test:

```
DROP SCHEMA public CASCADE ;
 CREATE SCHEMA public ;
 CREATE TABLE table_f2_f3 ( f2 INTEGER , f3 INTEGER  ) ;
 CREATE TABLE table_f3_f4_f5 ( f3 INTEGER , f4 INTEGER , f5 INTEGER  ) ;
 CREATE TABLE table_f4_f5_f6 ( f4 INTEGER , f5 INTEGER , f6 INTEGER  ) ;
 INSERT INTO table_f2_f3 VALUES ( 0 , 0 ) ;
 INSERT INTO table_f2_f3 VALUES ( 0 , 1 ) ;
 INSERT INTO table_f2_f3 VALUES ( NULL , 0 ) ;
 INSERT INTO table_f3_f4_f5 VALUES ( 0 , 1 , NULL ) ;
 INSERT INTO table_f3_f4_f5 VALUES ( 1 , 1 , 0 ) ;
 INSERT INTO table_f3_f4_f5 VALUES ( 1 , NULL , 0 ) ;
 INSERT INTO table_f4_f5_f6 VALUES ( 0 , 0 , 1 ) ;
 INSERT INTO table_f4_f5_f6 VALUES ( 1 , 0 , 0 ) ;
 INSERT INTO table_f4_f5_f6 VALUES ( NULL , 0 , 1 ) ;
 INSERT INTO table_f4_f5_f6 VALUES ( NULL , 1 , 0 ) ;
 INSERT INTO table_f4_f5_f6 VALUES ( NULL , 1 , 0 ) ;
SELECT * FROM table_f2_f3  JOIN ( table_f3_f4_f5  JOIN table_f4_f5_f6  USING ( f5  )  )  USING ( f3  )
 WHERE f6  IS  NULL  OR f3  >= f6  AND f6  <=  10000   ;
```

This branch returns the correct 6 rows (as confirmed by Postgres) rather than the incorrect 4 rows of the prior commit. The Explains are identical but there is a subtle difference in `EXPLAIN PHYSICAL PLAN`:

```
--- 6875.out	2022-11-21 13:58:50.351573735 +0100
+++ 16875.out	2022-11-21 13:58:50.391574707 +0100
...
 Source materialize.public.table_f2_f3
   filter=((#1) IS NOT NULL)
 Source materialize.public.table_f3_f4_f5
   filter=((#0) IS NOT NULL AND (#2) IS NOT NULL)
 Source materialize.public.table_f4_f5_f6
-  filter=((#1) IS NOT NULL AND ((#2) IS NULL OR (#2 <= 10000)))
+  filter=((#1) IS NOT NULL AND ((#2) IS NULL OR (#0 <= 10000)))
``` 

In other words, the predicate ` f6  <=  10000 ` was applied to the 1st column of ` Source materialize.public.table_f4_f5_f6` even though column `f6` is the third column in that source.

**Comment by ggevay** [comment]
Re Item No 1.:
I think the earlier plan is incorrect: The `(#0 = 0) AND (#1 = 0)` seems wrong: in the SQL we have `= 0` only on one field, and I don't see any correct reason that this would propagate to some other column (e.g., we don't have equality predicates between columns). A guess for what happened in the earlier plan is that ColumnKnowledge propagated the `= 0` upwards, and then the buggy predicate pushdown pushed an incorrect version back down.

**Review by philip-stoev — APPROVED** [review]
After several hours beating on this:

- no wrong results specific to this branch, that is, Postgres is in agreement with this branch as to the correct output and the prior commit is in the minority.
- no panics specific to this branch
- the dozens of explain plan and result differences that I examined appear to all fall into the patterns already in the comments. If you like, push any of  those in the SLT test if you think they provide any additional value.
Thanks.

**Comment by ggevay** [comment]
Re Item No 2.:
The earlier plan seems incorrect: `#1 <= (#1 + 123456789)` refers to the same field twice, while in the SQL this predicate refers to two different fields. The bug is occurring when this predicate is being pushed into `l1` and `l0`.

**Comment by ggevay** [comment]
Re Item No 3.:

> The Explains are identical

I ran the explains before and after the commit, and I see the same difference in the EXPLAINs as in the EXPLAIN PHYSICAL PLANs.

EXPLAIN after:
```
                            Optimized Plan                             
-----------------------------------------------------------------------
 Explained Query:                                                     +
   Project (#1, #0, #4, #3, #5, #7)                                   +
     Filter ((#7) IS NULL OR ((#7 <= 10000) AND (#1 >= #7)))          +
       Join on=(#1 = #2 AND #4 = #6) type=differential                +
         ArrangeBy keys=[[#1]]                                        +
           Filter (#1) IS NOT NULL                                    +
             Get materialize.public.table_f2_f3                       +
         ArrangeBy keys=[[#2]]                                        +
           Filter (#0) IS NOT NULL AND (#2) IS NOT NULL               +
             Get materialize.public.table_f3_f4_f5                    +
         ArrangeBy keys=[[#1]]                                        +
           Filter (#1) IS NOT NULL AND ((#2) IS NULL OR (#2 <= 10000))+
             Get materialize.public.table_f4_f5_f6                    +
                                                                      +
 Source materialize.public.table_f2_f3                                +
   filter=((#1) IS NOT NULL)                                          +
 Source materialize.public.table_f3_f4_f5                             +
   filter=((#0) IS NOT NULL AND (#2) IS NOT NULL)                     +
 Source materialize.public.table_f4_f5_f6                             +
   filter=((#1) IS NOT NULL AND ((#2) IS NULL OR (#2 <= 10000)))      +
```

EXPLAIN before:
```
                            Optimized Plan                             
-----------------------------------------------------------------------
 Explained Query:                                                     +
   Project (#1, #0, #4, #3, #5, #7)                                   +
     Filter ((#7) IS NULL OR ((#7 <= 10000) AND (#1 >= #7)))          +
       Join on=(#1 = #2 AND #4 = #6) type=differential                +
         ArrangeBy keys=[[#1]]                                        +
           Filter (#1) IS NOT NULL                                    +
             Get materialize.public.table_f2_f3                       +
         ArrangeBy keys=[[#2]]                                        +
           Filter (#0) IS NOT NULL AND (#2) IS NOT NULL               +
             Get materialize.public.table_f3_f4_f5                    +
         ArrangeBy keys=[[#1]]                                        +
           Filter (#1) IS NOT NULL AND ((#2) IS NULL OR (#0 <= 10000))+
             Get materialize.public.table_f4_f5_f6                    +
                                                                      +
 Source materialize.public.table_f2_f3                                +
   filter=((#1) IS NOT NULL)                                          +
 Source materialize.public.table_f3_f4_f5                             +
   filter=((#0) IS NOT NULL AND (#2) IS NOT NULL)                     +
 Source materialize.public.table_f4_f5_f6                             +
   filter=((#1) IS NOT NULL AND ((#2) IS NULL OR (#0 <= 10000)))      +
```

**Review by antiguru — APPROVED** [review]
Thanks, this looks good!
